Enable unused_crate_dependencies Rust lint, remove unused dependencies

findepi opened the pull request:
# Which issue does this PR close?

Fixes https://github.com/apache/arrow-rs/issues/6796

# Rationale for this change
 
Removing unused dependencies is expected to improve compile times.

# What changes are included in this PR?

- enable `unused_crate_dependencies` Rust lint for all crates
- remove unused dependencies flagged by that lint

# Are there any user-facing changes?

no

findepi commented:
The PR is expected to fail because (1) i didn't remove all the dependencies flagged by the lint and (2) there are false-positives. 
However, it failed in an unexpected way https://github.com/apache/arrow-rs/actions/runs/12027104363/job/33527467696?pr=6804

```
Run cargo clippy -- -D warnings
  cargo clippy -- -D warnings
  shell: sh -e {0}
  env:
    RUSTFLAGS: -C debuginfo=1
    RUST_BACKTRACE: 1
error: failed to parse manifest at `/__w/arrow-rs/arrow-rs/object_store/Cargo.toml`

Caused by:
  error inheriting `lints` from workspace root manifest's `workspace.lints`

Caused by:
  failed to find a workspace root
```

Any idea why `failed to find a workspace root`? i am not getting this error when i run
```
cargo clean
RUSTFLAGS="-C debuginfo=1" RUST_BACKTRACE=1 cargo +stable clippy -- -D warnings
```

locally

tustvold commented:
> Any idea why failed to find a workspace root? i am not getting this error when i run

object_store is not a member of the crate workspace as it is released separately

findepi commented:
Thanks, that explains why it fails on CI.
I missed the `working-directory: object_store` when replying CI steps locally.

tustvold commented:
Marking this as a draft as there appear to still be unresolved issues

findepi commented:
I changed the approach. Instead of enabling lint in the cargo.toml, i switched to enaling it on specific cargo invocations as part of CI. Why? Because it feels cargo lacks necessary information to make this check work well in all cases. It apparently does not distinguishes dependencies when compiling tests and examples, or tests with different feature flags. I couldn't find a way to make it work, without adding a lot of suppressions (`#allow` or fake imports). Moreover, being very strict about dev/test dependencies isn't probably useful, as it doesn't benefit Arrow's users in any way.

findepi commented:
cc @alamb

alamb commented on .github/workflows/arrow.yml:
```diff
@@ -164,36 +164,138 @@ jobs:
+          cargo clippy -p "$mod" -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --all-features -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --no-default-features -- -D unused_crate_dependencies
+      - name: Clippy arrow-schema
```
Nit- the names are slightly inconsistent:

```suggestion
      - name: Clippy arrow-schema with all features
```

findepi replied:
I'd go opposite direction, ie remove "with all features" as now this tests various combinations: "all features", "default features", "no features"

findepi replied:
done

alamb commented on .github/workflows/arrow.yml:
```diff
@@ -164,36 +164,138 @@ jobs:
+          cargo clippy -p "$mod" -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --all-features -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --no-default-features -- -D unused_crate_dependencies
+      - name: Clippy arrow-array
```
```suggestion
      - name: Clippy arrow-array with all features
```

findepi replied:
see https://github.com/apache/arrow-rs/pull/6804#discussion_r1879718116

alamb commented on arrow-flight/Cargo.toml:
```diff
@@ -43,11 +43,11 @@ base64 = { version = "0.22", default-features = false, features = ["std"] }
 # For Timestamp type
 prost-types = { version = "0.13.1", default-features = false }
-tokio = { version = "1.0", default-features = false, features = ["macros", "rt", "rt-multi-thread"] }
+tokio = { version = "1.0", default-features = false, features = ["macros", "rt", "rt-multi-thread"], optional = true }
```
💯

alamb reviewed (approved):
Thank you @findepi 

Sorry for the delay in reviewing

I reviewed the CI job output and it looks good to me (all the jobs ran well):
https://github.com/apache/arrow-rs/actions/runs/12116683077/job/33777545429?pr=6804

alamb commented:
I'll plan to merge this tomorrow unless there are any other comments. 

I hope to make the RC around Thursday

alamb commented:
Thanks again @findepi
